portfolio/src/main/java/com/google/sps/data/CuisineDatastore.java

```diff
@@ -0,0 +1,56 @@
+package com.google.sps.data;
+
+import java.util.Map;
+import com.google.appengine.api.datastore.*;
+import com.google.common.collect.ImmutableMap;
+
+/** Class that handles writing/reading cuisine data from the datastore. */
+public final class CuisineDatastore {
+    
+    /** Method that retreives cuisine entities from the datastore and formats it. */
+    public static Map<String, Integer> fetchCuisine() {
+      Query query = new Query("Cuisine");
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(query);
+      
+      ImmutableMap.Builder<String, Integer> cuisineVotes = ImmutableMap.builder();
+      for (Entity entity : results.asIterable()) {
+        String cuisine = (String) entity.getProperty("cuisine");
+        Long votes = (Long) entity.getProperty("votes");
+        cuisineVotes.put(cuisine, votes.intValue());
+      }
+
+      return cuisineVotes.build();
+    }
+
+    /** Method that updates the cuisine vote in the datastore.
+    Method throws {@link IllegalArgumentException} if the cuisine is empty or null */
+    public static void updateCuisine(String cuisine) {
+      if (cuisine.isEmpty() || cuisine == null) {
+        throw new IllegalArgumentException("Cuisine should not be empty or null");
+      }
+
+      Query.Filter filter = new Query.FilterPredicate("cuisine",
+        Query.FilterOperator.EQUAL, cuisine);
+      Query query = new Query("Cuisine").setFilter(filter);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(query);
+
+      boolean updated = false;
+
+      for (Entity entity : results.asIterable()) {
+        Integer votes = ((Long) entity.getProperty("votes")).intValue();
```

Comment:
Note that your code is not thread safe: if multiple users are calling your API at the same time you might skip a few counts.

I suppose there is no increment datastore query?

Reply:
Another way to solve the problem (if there is no increment) is to store the list of users that voted for it : (cuisine, userId), then to get the number of votes you count the number of these entries. That also prevents people from voting more than once and it's concurrency-safe since only the current user can update it's own entries and it's always setting a value not incrementing it.

Reply:
note: if you don't have a user Id you can generate a basic one in javascript and use a cookie or something like that to propagate it to the server user Ids should normally come from server but this is a cheap way of doing it.